validate-deployment-readiness

```diff
@@ -1,77 +1,106 @@
-#   requests to the API, with the key "lastMigrationHash"
-# - The `migrations` directory contains all migrations, alphabetically by date,
-#   with the exception of the `_template` file.
-#
-# Note: This incarnation of the script was built for cala-api specifically. If
-# we want to reuse this elsewhere, it'll need some modification.
-
-set -euo pipefail
-
-c_green=$'\033[0;32m'
-c_red=$'\033[0;31m'
-c_yellow=$'\033[1;33m'
-c_reset=$'\033[0m'
-
-main() {
-  deployment_target=${1:-other}
-  migrations=${2:-migrations}
-
-  case "$deployment_target" in
-    'staging')
-      status_url='https://api-stg.ca.la'
-      ;;
-    'production')
-      status_url='https://api.ca.la'
-      ;;
-    *)
-      echo 'Usage: validate-deployment-readiness [staging|production] [path_to_migrations]'
-      exit 1
-      ;;
-  esac
-
-  # The HTTP response body, followed by the response code on a new line
-  response_and_status=$(curl -s -w "\n%{http_code}" -o - -A 'validate-deployment-readiness script' $status_url)
-  status=$(echo "$response_and_status" | tail -n 1)
-
-  if [[ $status != '200' ]] ; then
-    # Assume that the server has crashed, so the best option is to _allow_ deployment
-    echo $c_yellow'Warning: Bad response status ('$status')!'$c_reset
-    exit 0
-  fi
-
-  deployed_hash=$(
-    echo $response_and_status |
-      egrep -o '"lastMigrationHash":"[^"]*"' |
-      cut -d '"' -f 4
-  )
-
-  echo $c_yellow'Deployed hash: '$c_reset$deployed_hash
-
-  latest_local_migration=$(
-    ls -Ar $migrations/*.js | # List all migrations, descending order by name
-      head -n 1 |  # Take the first one
-      tr -d '\n' | # Remove trailing newline
-      awk -F / '{print $NF}' # Filter down to the last field after a '/' character,
-                             # e.g. foo/bar/x.js becomes x.js
-  )
-
-  local_hash=$(echo -n $latest_local_migration | shasum -a 256 | cut -d ' ' -f 1)
-  echo $c_yellow'Local hash: '$c_reset$local_hash
-
-  if [[ $deployed_hash != $local_hash ]] ; then
-    echo $c_red'Hash mismatch! Make sure to run migrations _before_ deploying. Aborting.'$c_reset
-    exit 1
-  fi
-
-  echo $c_green'Migration hashes match. Ready to roll.'$c_reset
+#!/usr/bin/env node
+
+const process = require('process');
+const https = require('https');
+const fs = require('fs');
+const crypto = require('crypto');
+
+const green = '\033[0;32m';
+const red = '\033[0;31m';
+const yellow = '\033[1;33m';
+const reset = '\033[0m';
+
+const hash = crypto.createHash('sha256');
+
+main()
+  .then(() => process.exit(0))
+  .catch((err) => {
+    console.error(err);
+    process.exit(1);
+  });
+
+async function main() {
+  const [_, __, target, migrationPath] = process.argv;
+
+  if (!target || !migrationPath) {
+    console.log('Usage: validate-deployment-readiness [staging|production] [path_to_migrations]');
+    process.exit(1);
+  }
+
+  let deployedHash;
+  try {
+    deployedHash = await getDeployedHash(target);
+  } catch (error) {
```

Comment:
Not totally convinced on this one — we should probably only be "failing safe" here if the response status was non-2XX or the body wasn't JSON, not for any error at all.

Reply:
Good call.